### PR TITLE
[inductor][estimations] Support sparsity_hint for flex_attention flop estimates

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #182537

flex_attention with sparse masks (e.g., causal, block-sparse) does
significantly fewer FLOPs than the dense O(N^2) upper bound. This
causes the roofline cost model to overestimate compute time, which
in turn skews overlap scheduling and SAC decisions.

This PR threads a `sparsity_hint` annotation (0.0-1.0 fraction of
skipped work) from `torch.fx.traceback.annotate` through the flop
counting infrastructure so that flex_attention forward and backward
flop formulas can scale down their estimates accordingly.

The plumbing:
- `get_compute_time` accepts an optional `node_meta` dict, forwarded
  as `_node_meta` kwarg to flop formulas
- `estimate_roofline_runtime_ms` passes `node.meta` through
- flex_attention flop formulas read `sparsity_hint` from `_node_meta`
- `FlexAttentionAutogradOp` captures `sparsity_hint` in forward and
  re-annotates the backward call so both directions benefit

Authored with Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/test_flop_counter.py [hunk 1]
```diff
@@ -1344,6 +1344,82 @@ def test_flex_attention_roofline_estimate(self):
...
+                return flex_attention(q, k, v)
+
+        class Sparse(torch.nn.Module):
+            def forward(self, q, k, v):
+                with annotate({"sparsity_hint": 0.9}):
+                    return flex_attention(q, k, v)
+
+        # Shapes must be large enough that the roofline estimate is compute-bound
+        # (not memory-bound), so sparsity_hint actually reduces the estimate.
+        q = torch.randn(1, 32, 2048, 128, device="cuda", dtype=torch.bfloat16)
```
Comment (drisspg) [inline test/test_flop_counter.py]: could we use fake tensors for this?

File: torch/_higher_order_ops/flex_attention.py [hunk 2]
```diff
@@ -849,41 +853,47 @@ def backward(  # type: ignore[override]
...
-                full_q_indices,
-                Q_BLOCK_SIZE,
-                KV_BLOCK_SIZE,
-                mask_graph,
-            ),
-            scale,
-            kernel_options,
-            score_mod_other_buffers,
-            mask_mod_other_buffers,
+        _sparsity_ctx = (
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:856]: why do we need this here?

Comment (IvanKobzarev) [reply torch/_higher_order_ops/flex_attention.py:856]: This is propagation to backward, that flex_backward will have the same "sparsity" for estimations

File: test/test_flop_counter.py (hunk 1 above)
Comment (IvanKobzarev) [reply test/test_flop_counter.py]: good catch, yes we can. will convert.

File: torch/_higher_order_ops/flex_attention.py (hunk 2 above)
Comment (drisspg) [reply torch/_higher_order_ops/flex_attention.py:856]: does aotautograd not preserve annotations?

Comment (IvanKobzarev) [reply torch/_higher_order_ops/flex_attention.py:856]: no, aot_autograd generates backward via tracing autograd.grad, so the fx.nodes will be completely new. This registration for backward for hop.flex will be inserting the node and we add meta.

**Review by drisspg — APPROVED** [review]
